proposals/0000-tuple-syntax.rst

```diff
@@ -0,0 +1,324 @@
+=================================
+
+.. author:: Richard Eisenberg (with much influence from collaborators)
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header::  This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/475>`_.
+.. contents::
+.. sectnum::
+
+Introduce new names for the list type and tuple types to allow for users
+to write pun-free code. In addition, add a new extension that disables
+the built-in syntax for these types, thus requiring users to use the new
+names.
+
+.. _`#270`: https://github.com/ghc-proposals/ghc-proposals/pull/270
+.. _`#281`: https://github.com/ghc-proposals/ghc-proposals/pull/281
+.. _`#378`: https://github.com/ghc-proposals/ghc-proposals/pull/378
+.. _`#473`: https://github.com/ghc-proposals/ghc-proposals/pull/473
+
+Motivation
+----------
+Proposals `#270`_, `#281`_, `#378`_, and `#473`_ all struggle when faced
+with *puns*, the use of an identifier that can be interpreted either in the
+type-level namespace or the term-level namespace (but with different meanings).
+
+Furthermore, puns can be hard for newcomers to Haskell. It is not hard to jump
+from ::
+
+  single :: [Int]
+  single = [5]
+
+to ::
+
+  just :: Maybe Int
+  just = Maybe 5
+
+even though the latter is nonsense.
+
+Puns also cause trouble when describing what we are talking about: conversations
+around these features invariably require people to say "``[]``-the-type" or
+"``[]``-the-empty-list" or similar.
+
+This proposal thus allows for a common way to refer to tuples and lists without
+punning. It is entirely opt-in, though we expect the new phrasing to become popular
+over time.
+
+Proposed Change Specification
+-----------------------------
+
+1. Add the following definitions to a new module ``GHC.Tuple.Prim``::
+
+     data Unit = ()
+     data Solo a = MkSolo a    -- this is a change from today's `data Solo a = Solo a`
+
+     type Tuple0 = Unit
+     type Tuple1 = Solo
+     data Tuple2 a b = (a, b)
+     data Tuple3 a b c = (a, b, c)
+     -- ...
+     data Tuple64 ... = (...)
+
+#. Add the following definitions to ``GHC.Classes`` (re-exported from ``GHC.Exts``). These
+   replace similar definitions today::
+
+     class CUnit
+     instance CUnit
+
+     class a => CSolo a
+     instance a => CSolo a
+
+     type CTuple0 = CUnit
+     type CTuple1 = CSolo
+
+     class (a, b) => CTuple2 a b
+     instance (a, b) => CTuple2 a b
+
+     -- ...
+
+     class (...) => CTuple64 ...
+     instance (...) => CTuple64 ...
+
+#. Remove existing tuple definitions from ``GHC.Tuple``.
+
+#. Add the following definitions to ``GHC.Tuple``, which re-exports all of ``GHC.Tuple.Prim``::
+
+     type TupleNKind :: Nat -> Type     -- Nat is from GHC.TypeLits
+     type family TupleNKind n = r | r -> n where
+       TupleNKind 0 = Type
+       TupleNKind n = Type -> TupleNKind (n-1)   -- this fails the injectivity check, but a little magic will allow this
```

Comment:
Do you envision the magic to be specific to `TupleNKind` or all definition that try to use `Nat` inductively?

Reply:
I think we should do what's necessary to get TupleNKind to compile. If we can do so generally, then all the better.

Right now, GHC has some built-in magic so that, if `n0-1 ~ m` and `n1-1 ~ m`, then GHC will know that `n0 ~ n1`. Note that this magic isn't available to non-built-in functions: it's using the fact that both calls to `-` take the same second argument. But my guess is that this particular existent magic isn't consulted during the injectivity check. So, either we consult it during the injectivity check, or we whip up some custom magic for `TupleNKind`. The former is preferable, but the latter is acceptable. But I don't wish to hold up this proposal on, say, a specification of exactly how to do this.